### PR TITLE
Add Burn msg to erc20 contract

### PR DESCRIPTION
**Pull request by anilCSE**

### CODE REVIEW

File: erc20/src/contract.rs [hunk 1]
```diff
@@ -211,6 +212,57 @@ fn try_approve<S: Storage, A: Api>(
     Ok(res)
 }
 
+/**
+* Destroy tokens
+*
+* Remove `amount` tokens from the system irreversibly
+*
+* @param amount the amount of money to burn
+*/
```
Comment (webmaster128) [inline erc20/src/contract.rs]: Rust uses triple slash doc comments, like

```rust
/// Destroy tokens
///
/// Remove `amount` tokens from the system irreversibly
...
```

Comment (webmaster128) [reply erc20/src/contract.rs]: I'd mention who's balance is reduced by the burning

**Comment by anilCSE** [comment]
> Very nice, I like. Let @ethanfrey have a look as well.
> 
> Can you rebase on top of current master? ` erc20/contract.wasm` and `erc20/hash.txt` were updated recently on master, which is causing the conflicts.

Updated! Thank you for a quick review.

**Comment by webmaster128** [comment]
Right now the CI fails because of a missing `cargo fmt && cargo schema`

**Comment by anilCSE** [comment]
> Right now the CI fails because of a missing `cargo fmt && cargo schema`

Updated

File: erc20/src/contract.rs [hunk 2]
```diff
@@ -211,6 +212,57 @@ fn try_approve<S: Storage, A: Api>(
...
+    if account_balance < amount_raw {
+        return dyn_contract_err(format!(
+            "insufficient funds to burn: balance={}, required={}",
+            account_balance, amount_raw
+        ));
+    }
+    account_balance -= amount_raw;
+    balances_store.set(owner_address_raw.as_slice(), &account_balance.to_be_bytes());
+
+    let mut config_store = PrefixedStorage::new(PREFIX_CONFIG, &mut deps.storage);
```
Comment (ethanfrey) [inline erc20/src/contract.rs]: Good, update total supply as well.

File: erc20/src/tests.rs [hunk 3]
```diff
@@ -1033,6 +1033,167 @@ mod transfer_from {
...
+        ); // -1
+        assert_eq!(
+            get_balance(&deps.api, &deps.storage, &HumanAddr("addr1111".to_string())),
+            22
+        );
+        assert_eq!(get_total_supply(&deps.storage), 32);
+    }
+
+    #[test]
+    fn can_burn_zero_amount() {
```
Comment (ethanfrey) [inline erc20/src/tests.rs]: I would add a test of burning from an account with no balance. (should error)

Comment (anilCSE) [reply erc20/src/tests.rs]: I think `fails_on_insufficient_balance` covers that right? I can add one but looks duplicate

Comment (anilCSE) [reply erc20/src/tests.rs]: May be we can instead have validation for `amount` and it should be > 0

Comment (webmaster128) [reply erc20/src/tests.rs]: Hmm... I tend to agree with @anilCSE. However, you could treat those cases as different things:
- a user of the contract has insufficient balance, but is known to the contract
- an address that was never seen by the contract before wants to burn

However, this woud test _get balance of address_ more than the burn.

File: erc20/src/contract.rs [hunk 4]
```diff
@@ -211,6 +212,57 @@ fn try_approve<S: Storage, A: Api>(
...
+fn try_burn<S: Storage, A: Api>(
+    deps: &mut Extern<S, A>,
+    env: Env,
+    amount: &str,
+) -> Result<Response> {
+    let owner_address_raw = &env.message.signer;
+    let amount_raw = parse_u128(amount)?;
+
+    let mut balances_store = PrefixedStorage::new(PREFIX_BALANCES, &mut deps.storage);
+    let mut account_balance = read_u128(&balances_store, owner_address_raw.as_slice())?;
```
Comment (webmaster128) [inline erc20/src/contract.rs]: Wouldn't it make sense to use `read_balance` here? Then we could unit test this to ensure a non existent addresses leads to 0.

File: erc20/src/tests.rs (hunk 3 above)
Comment (webmaster128) [reply erc20/src/tests.rs]: > May be we can instead have validation for `amount` and it should be > 0.
> Burning zero tokens doesn't make sense. It's unnecessary gas cost, we should just throw error there

I would avoid adding unnecessary code branches. There is no requirement to disallow burning 0. You can send 0 as well.

Comment (anilCSE) [reply erc20/src/tests.rs]: Agreed

File: erc20/src/contract.rs (hunk 4 above)
Comment (anilCSE) [reply erc20/src/contract.rs]: Why `read_balance` is returning `Result<u128> `and why not just `u128`? I believe read_balance should handle the errors and return just `0`. If we need to validate account address as well, that should be done outside of `read_balance` imo.

File: erc20/src/contract.rs [hunk 5]
```diff
@@ -225,8 +225,7 @@ fn try_burn<S: Storage, A: Api>(
     let owner_address_raw = &env.message.signer;
     let amount_raw = parse_u128(amount)?;
 
-    let mut balances_store = PrefixedStorage::new(PREFIX_BALANCES, &mut deps.storage);
-    let mut account_balance = read_u128(&balances_store, owner_address_raw.as_slice())?;
+    let mut account_balance = read_balance(&mut deps.storage, owner_address_raw).unwrap();
```
Comment (ethanfrey) [inline erc20/src/contract.rs]: Don't unwrap, return an error.

Replace `.unwrap()` with `?`. Works like magic.

**Pull request merged**